docs/proposals/kubectl-login.md

```diff
@@ -0,0 +1,271 @@
+can be supplied without user interaction (refreshing an OAuth token, redeeming
+a Kerberos ticket, etc.). Users with fully configured kubeconfigs will only see
+the user they're logged in as, useful for opaque credentials such as X509 certs
+or bearer tokens.
+
+The command differs from `kubectl config` by:
+
+* Communicating with the API Server to determine if the user is supplying valid auth events.
+* Validating input and being opinionated about the input it asks for.
+* Triggering authentication challenges for example:
+  * Basic auth: Actually try to communicate with the API server.
+  * OpenID Connect: Create an OAuth2 redirect.
+
+However `kubectl login` should still be seen as a supplement to, not a
+replacement for, `kubectl config` by helping validate any kubeconfig generated
+by the latter command.
+
+## Server endpoints
+
+On the API Server side this command requires two non-resource paths.  First a
+path to discover the authentication strategies supported by the API Server.
+
+> GET /auth/authenticators
+
+```
+{
+  "kind": "AuthenticationModes",
+  "apiVersion": "authentication.k8s.io/v1beta1",
+  "authenticators": [
+    {
+      "mode": "tokenfile"
+    },
+    {
+      "mode": "oidc",
+      "config": {
+        "issuer": "https://accounts.google.com",
+        "trusted-client-id": "foobar"
+      }
+    },
+    {
+      "mode": "keystone",
+      "config": {
+        "url": "https://keystone.example.com"
+      }
+    },
+  ]
+}
+```
+
+And another path to get the user a request is logged in as.
+
+> GET /auth/whoami
+
+```
+{
+  "kind": "UserInfo",
+  "apiVersion": "authentication.k8s.io/v1beta1",
+  "nameuser": "janedoe",
+  "groups": [
+    "admins",
+    "kube-system"
+  ]
+}
+```
+
+To support these endpoints, add a top level type to `pkg/apis/authentication.k8s.io`
+and reuse the existing `UserInfo` type from that API Group.
+
+```go
+// AuthenticationModes lists the authentication strategies available, to allow
+// clients to discover which strategies are valid.
+type AuthenticationModes struct {
+    unversioned.TypeMeta `json:",inline"`
+
+	// The list of authentication strategies enabled on this API server.
+	Authenticators []Authenticator `json:"authenticators"`
+}
+
+
+// Authenticator is a strategy for authenticating with the Kubernetes API.
+type Authenticator struct {
+	// The type of the authenticator such as "tokenfile" or "webhook".
+	Mode string `json:"mode"`
+
+	// Config contains public fields which indicate the configuration of the plugin
+	// to help clients auto-fill required fields.
+	Config map[string]string `json:"config"`
+}
+```
+
+## Examples
+
+If kubeconfig isn't configured, `kubectl login` will attempt to fully configure
+and validate the client's credentials.
+
+```
+$ kubectl login
+Cluster URL []: https://172.17.4.99:443
+Cluster CA [*omit to use host certs*]: ${PWD}/ssl/ca.pem
+Cluster name ["kubernetes"]:
+
+The kubernetes server supports the following methods:
+
+  1. Bearer token
```

Comment:
I suppose that's true, but I'd like to make sure that this is tied into the specific `Authenticator` rather than a random map somewhere in `kubectl`  So perhaps instead of tying this info to the API object we need to add something to the `Authenticator` interface?